### PR TITLE
[Bugfix] Detach Loaders after running entrypoint

### LINKED ISSUE #8868: Predicting with custom dataloader overwrites "predict_dataloader()" method of module
## 🐛 Bug

When predicting using trainer.predict(module,DataLoader(my_dataset)), the trainer will overwrite the "predict_dataloader()" method of "module".

### To Reproduce

```
import torch
from torch.utils.data import Dataset, DataLoader, TensorDataset
from torch.nn import MSELoss, Linear, ReLU
import pytorch_lightning as pl
from pytorch_lightning.loggers import NeptuneLogger



class TestDataset(Dataset):
    def __init__(self, length = 100):
        self.length = length

    def __len__(self):
        return self.length

    def __getitem__(self, item):
        return torch.tensor(item).float(), torch.tensor(item**2).float()

tensor_dataset = TensorDataset(torch.tensor(torch.linspace(1,100)),
                               torch.tensor(torch.linspace(1,100**2)))

class Module(pl.LightningModule):
    def __init__(self):
        super(Module,self).__init__()
        self.criterion = MSELoss()
        self.dataset = TestDataset()
        self.layers = torch.nn.Sequential(Linear(1, 5), ReLU(),
                                          Linear(5, 5), ReLU(),
                                          Linear(5, 5), ReLU(),
                                          Linear(5, 1))

    def forward(self, x):
        return self.layers(self.layers(x.view(x.shape[0], -1)))

    def training_step(self, batch, batch_idx):
        x, y = batch
        train_loss = self.criterion(self(x), y)
        self.log("train_loss", train_loss, on_step=False, on_epoch=True)
        return train_loss

    def predict_step(self, batch, batch_idx, dataloader_ix):
        x, y = batch
        train_loss = self.criterion(self(x), y)
        return train_loss

    def train_dataloader(self):
        return DataLoader(self.dataset,batch_size=10)

    def predict_dataloader(self):
        return DataLoader(self.dataset, batch_size=10)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters())

trainer = pl.Trainer( gpus=[2])
module = Module()
trainer.predict(module)
print(" ")
print("This should give TestDataset class:")
print(" ")
print(module.predict_dataloader().dataset)
trainer.predict(module, DataLoader(tensor_dataset))
print(" ")
print("This should give TestDataset class:")
print(module.predict_dataloader().dataset)
```

### Expected behavior

The DataLoader returned from "predict_dataloader()" should be the DataLoader specified in the modules function, not the dataloader passed to trainer.predict() at some point in time before that.

### Environment

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
                - NVIDIA GeForce RTX 2080 Ti
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.20.2
        - pyTorch_debug:     False
        - pyTorch_version:   1.8.1
        - pytorch-lightning: 1.3.5
        - tqdm:              4.61.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #71~20.04.1-Ubuntu SMP Thu Jul 15 17:46:08 UTC 2021

### Additional context

I need to run trainer.predict() in two separate steps, because I need to do something with the returned data in between. I cant pass it as a list of dataloaders.

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #8868 

Restores the original dataloaders after the entrypoint if replaced during the run.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -536,6 +536,8 @@ def fit(
 
         self._run(model)
 
+        self.data_connector.detach_data(model)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:539]: Why not put these calls inside `_run` if all call it?

**Comment by justusschock** [comment]
@ananthsub This is a short-term fix for the issue, which still does not resolve #6030

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/trainer.py:539]: I thought about that, but since attaching also happens outside of run, I chose to have it on the same level. can also move it though

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:539]: i think it makes sense to include in `_run` so it's shared. specifically, does this belong in the teardown hook?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/trainer/trainer.py]: moved to teardown

File: tests/trainer/test_data_loading.py [hunk 2]
```diff
@@ -254,3 +254,62 @@ class CustomSampler(Sampler):
...
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
+
+    trainer.test(model, loader)
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
```
Comment (carmocca) [inline tests/trainer/test_data_loading.py:315]: Would this be enough to test?

(i have omitted `validate` because it gets set with `fit`)

```suggestion
    trainer.fit(model, loader, loader)
    trainer.predict(model, loader)
    trainer.test(model, loader)

    assert len(model.train_dataloader()) == 64
    assert len(model.val_dataloader()) == 64
    assert len(model.predict_dataloader()) == 64
    assert len(model.test_dataloader()) == 64
```

File: pytorch_lightning/trainer/callback_hook.py [hunk 3]
```diff
@@ -52,6 +53,8 @@ def teardown(self, stage: Optional[str] = None) -> None:
         for callback in self.callbacks:
             callback.teardown(self, self.lightning_module, stage=stage)
 
+        self.data_connector.detach_data(self.lightning_module)
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/callback_hook.py:57]: This should be in `Trainer._call_teardown_hook`

File: tests/trainer/test_data_loading.py (hunk 2 above)
Comment (justusschock) [reply tests/trainer/test_data_loading.py:315]: This would still catch it, but I explicitly wanted to have tests for all the entrypoints that each of them resets all loaders. We can keep it that way though.

File: pytorch_lightning/trainer/callback_hook.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/trainer/callback_hook.py]: moved again :D

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 4]
```diff
@@ -117,19 +117,23 @@ def attach_dataloaders(
         # functions to overwrite with these implementations
         if train_dataloaders is not None:
             self.trainer.train_dataloader = None
-            model.train_dataloader = _PatchDataLoader(train_dataloaders)
+            train_dataloader = _PatchDataLoader(train_dataloaders, "train")
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/data_connector.py:120]: @justusschock do you think it is possible to avoid patching all together? There are good reasons not to do it #6030 from a user's perspective. While I believe your PR solves the major issue, there could still be problems when the user wants to call their dataloader method to produce a fresh dataloader, for example in a callback (nothing is stopping them from doing so).

Some work has started here #7522

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/data_connector.py:120]: @awaelchli @justusschock I think we need a centralization point for where dataloaders come from. Patching the dataloader methods onto the model is using the model as the source of truth, but the side effects are visible to the end user. Rather, we could create an internal DataHolder that could be used to pool the object that has the DataHooks available. This would also codify the priority/precedence across datamodules, lightning module, and dataloaders passed directly to the trainer.

It will also raise the question: what happens when both the datamodule and the lightning module have these hooks implemented? https://github.com/PyTorchLightning/pytorch-lightning/blob/e0605472306d6b95bf2616ab88f8c29f4498402e/pytorch_lightning/core/hooks.py#L455-L807

do we raise an exception? do we run only the datamodule's, if available? do we run both and if so, in what order?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/data_connector.py:120]: @ananthsub @awaelchli  let's continue this discussion in #6030 and merge this as a fix :)

I copied both your questions/statements there :)
